tests/Api/Shop/Checkout/CartTest.php

```diff
@@ -0,0 +1,524 @@
+        );
+
+        $this->assertResponseViolations(
+            $this->client->getResponse(),
+            [
+                ['propertyPath' => '', 'message' => 'An empty order cannot be completed.'],
+            ],
+        );
+    }
+
+    /** @test */
+    public function it_does_not_allow_update_without_require_billing_address(): void
+    {
+        $this->loadFixturesFromFiles([
+            'channel.yaml',
+            'cart.yaml',
+            'country.yaml',
+        ]);
+
+        $tokenValue = $this->pickUpCart();
+        $this->addItemToCart('MUG_BLUE', 3, $tokenValue);
+
+        $this->client->request(
+            method: 'PUT',
+            uri: sprintf('/api/v2/shop/orders/%s', $tokenValue),
+            server: self::CONTENT_TYPE_HEADER,
+            content: json_encode([
+                'email' => 'oliver@doe.com',
+                'shippingAddress' => [
+                    'firstName' => 'Oliver',
+                    'lastName' => 'Doe',
+                    'phoneNumber' => '123456789',
+                    'countryCode' => 'US',
+                    'provinceCode' => 'US-MI',
+                    'city' => 'New York',
+                    'street' => 'Broadway',
+                    'postcode' => '10001',
+                ],
+            ], \JSON_THROW_ON_ERROR),
+        );
+
+        $this->assertResponseViolations(
+            $this->client->getResponse(),
+            [
+                ['propertyPath' => '', 'message' => 'Please provide a billing address.'],
+            ],
+        );
+    }
+
+    /** @test */
+    public function it_does_not_allow_update_without_require_shipping_address(): void
+    {
+        $this->loadFixturesFromFiles([
+            'channel.yaml',
+            'cart.yaml',
+            'country.yaml',
+        ]);
+
+        $tokenValue = $this->pickUpCart(channelCode: 'MOBILE');
+        $this->addItemToCart('MUG_BLUE', 3, $tokenValue);
+
+        $this->client->request(
+            method: 'PUT',
+            uri: sprintf('/api/v2/shop/orders/%s', $tokenValue),
+            server: self::CONTENT_TYPE_HEADER,
+            content: json_encode([
+                'email' => 'oliver@doe.com',
+                'billingAddress' => [
+                    'firstName' => 'Oliver',
+                    'lastName' => 'Doe',
+                    'phoneNumber' => '123456789',
+                    'countryCode' => 'US',
+                    'provinceCode' => 'US-MI',
+                    'city' => 'New York',
+                    'street' => 'Broadway',
+                    'postcode' => '10001',
+                ],
+            ], \JSON_THROW_ON_ERROR),
+        );
+
+        $this->assertResponseViolations(
+            $this->client->getResponse(),
+            [
+                ['propertyPath' => '', 'message' => 'Please provide a shipping address.'],
+            ],
+        );
+    }
+
+    /** @test */
+    public function it_does_not_allow_update_without_invalid_data(): void
```

Comment:
```suggestion
    public function it_does_not_allow_update_with_both_invalid_and_missing_data(): void
```
Or something similar.